### ISSUE TITLE
Bad Ruby detection when multiple rubies installed on OS X

### ISSUE THREAD

**Issue by apjanke** [weechat/weechat#714]
## Summary

I ran in to a failure building Ruby support in to weechat 1.4 on OS X systems where there are multiple `ruby` installations. The build fails with a link error complaining about a missing `libruby` symbol if there is an incompatible set of Ruby installations at various locations.

```
[ 69%] Linking C shared module ruby.so
cd /tmp/weechat20160411-68061-18wusr/weechat-1.4/build/src/plugins/ruby && /usr/local/Cellar/cmake/3.5.1/bin/cmake -E cmake_link_script CMakeFiles/ruby.dir/link.txt --verbose=1
/usr/local/Library/ENV/4.3/clang    -Wall -Wextra -Werror-implicit-function-declaration -DNDEBUG -isysroot /Applications/Xcode-7.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.11 -bundle -Wl,-headerpad_max_install_names  -o ruby.so CMakeFiles/ruby.dir/weechat-ruby.o CMakeFiles/ruby.dir/weechat-ruby-api.o /usr/local/lib/libruby.dylib ../libweechat_plugins_scripts.a 
Undefined symbols for architecture x86_64:
  "_rb_funcall2", referenced from:
      _protect_funcall0 in weechat-ruby.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

If there is only the system `ruby` installed, the build works fine. Another user has also reported this happening with just a Homebrew-installed `ruby` under `/usr/local`, but I could not reproduce it myself without the Fink `ruby` also installed.
## Environment

A Mac Homebrew user [first encountered this](https://github.com/Homebrew/homebrew-core/issues/117). The build would fail for them if they had a version-suffixed Ruby (with a `ruby2.0` command) installed in `/sw` (from Fink) in addition to the system Ruby and Homebrew-installed Ruby. Tested on OS X 10.11.4.

I was able to reproduce it with a fresh Fink install yesterday.

To reproduce:
- Install Fink
- `fink install ruby`
- Install Mac Homebrew
- `brew install ruby`
- Get `/usr/local/bin` on your path ahead of `/usr/bin` and `/sw/bin`
- Build `weechat` with Ruby enabled
## Diagnosis

Seems like an issue with the [`FindRuby.cmake`](https://github.com/weechat/weechat/blob/24c035ae9dd3258257e610c7d3ce316468fe882e/cmake/FindRuby.cmake) implementation in weechat. It's picking up header paths from one Ruby, but the library paths from a different one.

```
-- checking for one of the modules 'ruby-2.2;ruby-2.1;ruby-2.0;ruby-1.9;ruby-1.8'
-- Found Ruby: /sw/bin/ruby2.0
-- RUBY_POSSIBLE_LIB_PATH: /sw/lib
-- RUBY_RUBY_LIB_PATH: /sw/lib/ruby/2.0
-- RUBY_LIB: /usr/local/lib/libruby.dylib
```

There's a couple things going on in that `FindRuby.cmake`.
- PATHS vs HINTS
- Separate library and program searches, with different versioned-binary name lists
#### Paths vs HINTS

In the `find_library()` call done after locating the `ruby` executable, it uses `PATHS` to pass the detected Ruby installation location along to the library search routine.

```
    find_library(RUBY_LIB
      NAMES ruby-1.9.3 ruby1.9.3 ruby193 ruby-1.9.2 ruby1.9.2 ruby192 ruby-1.9.1 ruby1.9.1 ruby191 ruby1.9 ruby19 ruby1.8 ruby18 ruby
      PATHS ${RUBY_POSSIBLE_LIB_PATH} ${RUBY_RUBY_LIB_PATH}
      )
```

It should probably be using `HINTS` instead of `PATHS` here. According to the CMake [`find_library` documentation](https://cmake.org/cmake/help/v3.4/command/find_library.html), `HINTS` is for "paths computed by system introspection, such as a hint provided by the location of another item already found", which is what we have here. `PATHS` is for hard-coded guesses, and is a lower-priority item. Specifically, `PATHS` is lower priority than `CMAKE_SYSTEM_*_PATHS`, which contains a lot of locations, including `/usr/local`, `/sw`, and `/opt/local` by default on OS X. `HINTS` is higher priority than `CMAKE_SYSTEM_*_PATHS`.
#### Separate library and program searches

The Ruby library is detected with a `find_library` call after `find_program` and the header detection. PATHS is used to pass along the earlier result, but that doesn't really constrain the search.

The two calls use different lists of version-suffixed names.

```
  find_program(RUBY_EXECUTABLE
    NAMES ruby2.3.0 ruby230 ruby2.3 ruby23 ruby2.2.3 ruby223 ruby2.2.2 ruby222 ruby2.2.1 ruby221 ruby2.2.0 ruby220 ruby2.2 ruby22 ruby2.1.7 ruby217 ruby2.1.6 ruby216 ruby2.1.5 ruby215 ruby2.1.4 ruby214 ruby2.1.3 ruby213 ruby2.1.2 ruby212 ruby2.1.1 ruby211 ruby2.1.0 ruby210 ruby2.1 ruby21 ruby2.0 ruby20 ruby1.9.3 ruby193 ruby1.9.2 ruby192 ruby1.9.1 ruby191 ruby1.9 ruby19 ruby1.8 ruby18 ruby
    PATHS /usr/bin /usr/local/bin /usr/pkg/bin
    )
...
    find_library(RUBY_LIB
      NAMES ruby-1.9.3 ruby1.9.3 ruby193 ruby-1.9.2 ruby1.9.2 ruby192 ruby-1.9.1 ruby1.9.1 ruby191 ruby1.9 ruby19 ruby1.8 ruby18 ruby
      PATHS ${RUBY_POSSIBLE_LIB_PATH} ${RUBY_RUBY_LIB_PATH}
      )
```

This means that if you have a Ruby 2.0 or newer, it might be detected by that `find_program`, but not by the `find_library`, which could end up falling through to an older version or plain `ruby`.
## In closing

It's important that the headers and lib files at least are drawn from the same Ruby installation. The CMake Ruby detection code may need to be altered to be more strict about that, enforcing that the library files are taken from the `ruby` found with `find_program` if at all possible.

CMake's [own `FindRuby.cmake`](https://github.com/Kitware/CMake/blob/9ac11bc25d6b9f5e7db786034f922d96613e6143/Modules/FindRuby.cmake) has been substantially updated since the weechat one was made. Maybe you could just use that instead?

This is not a high priority issue, since it only affects unusual configurations, and there are workarounds.

**Comment by megalithic** [comment]
👍 definitely having this issue with `chruby`-installed rubies. seems to have started with weechat 1.4 best I can tell.

**Comment by flashcode** [comment]
This should be fixed by issue #1721, the detection of Ruby has been simplified.
If not, please show the build output with latest devel version.
